src/hotspot/share/opto/parse2.cpp

```diff
@@ -99,6 +99,17 @@ void Parse::array_store(BasicType bt) {
   if (elemtype == TypeInt::BOOL) {
     bt = T_BOOLEAN;
   }
+
+  if (DoPartialEscapeAnalysis && bt == T_OBJECT) {
+    PEAState& as = jvms()->alloc_state();
+    VirtualState* obj_vs = as.as_virtual(val);    // obj is tracked and Virtual
+    VirtualState* ary_vs = as.as_virtual(array);  // array is not-tracked or Escaped
+
+    if (ary_vs == nullptr && obj_vs != nullptr) {
+      EscapedState* escaped = as.materialize(this, val);
```

Comment:
Can we consider this a bug fix? I'm surprised we have not run into this before. I guess raw array's are not used so much in Java code.

Reply:
yes, it's a bug. Prior to this patch,  we convert an object from Virtual to Escaped in the following 2 cases.
1. put-field and put-static-field (GlobalEscaped)
2. invoke bytecodes (ArgEscaped)

This is the 3rd case. Are there anymore cases?  
3. aatore (GlobalEscaped)

We also convert an object from Virtual to Escaped when we encounter a loop header, but we don't materialize it.

Reply:
@caojoshua one thing I don't quite understand here. Is it possible that `array` is null here? 

`array_store_check()` before performs subtype check, but I don't find explicit check that array is null. it feel that C2 assume it's impossible. do you know why?